### ISSUE TITLE
[FEATURE REQUEST] Allow masking/scaling of all HEMCO_Config.rc entries

### ISSUE THREAD

**Issue by lizziel** [geoschem/geos-chem#339]
This is part 2 of separate issue https://github.com/geoschem/geos-chem/issues/232 which is now closed with the pending release of GEOS-Chem 12.9.0.

Meteorology fields and other HEMCO fields retrieved in GeosCore subroutines all accessed HEMCO via subroutine HCO_GetPtr prior to GEOS-Chem 12.9.0. This subroutine retrieves a pointer to the data read and stored in HEMCO, but that data does not have masks or scaling configured in HEMCO_Config.rc applied. This means that applying masks or scaling in HEMCO_Config.rc will have no effect on the model if the data is retrieved in GeosCore.

To have masks and scaling applied you must change the call to HCO_GetPtr to instead call HCO_EvalFld. The USE statement for HCO_EvalFld must be added (USE HCO_Calc_Mod). HCO_EvalFld takes an array as input/output, unlike HCO_GetPtr which takes a pointer. Therefore you must add arrays to store the data and allocate and deallocate them locally. The array precision should be real flexible precision (fp), unlike the pointer passes to HCO_GetPtr which was single precision.

In GEOS-Chem 12.9.0 most calls to HCO_GetPtr were changed to HCO_EvalFld. This was done for all GeosCore files except the following:

strat_chem_mod.F90
get_met_mod.F90
olson_landmap_mod.F90
modis_lai_mod.F90
hcoi_gc_main_mod.F90
land_mercury_mod.F90
mercury_mod.F90
ocean_mercury_mod.F90
pops_mod.F90
rrtmg_rad_transfer_mod.F90

Work in strat_chem_mod.F90 is delayed pending decisions about retiring tropchem and how to handle the stratosphere. Work in get_met_mod.F90 and hcoi_gc_main_mod.F90 is delayed pending coordination with other future HEMCO updates, including updates related to meteorology input. Work in olson_landmap_mod.F90 and modis_lai_mod.F90 may not be something we want to do; scaling/masking the landmap and leaf area index is non-physical and it is better for users to create new input files. POPs, Hg, and RRTMG specialty simulation work is to be completed in the future.

This issue is to track progress of completion of these updates in GEOS-Chem. The work is anticipated to have no impact on the GEOS-Chem benchmark output.

# Request a new GEOS-Chem feature

### Please see our Github tutorial videos
If this is the first time you are submitting a GEOS-Chem issue via Github, we recommend that you first view our tutorial videos on [youtube.geos-chem.org](http://youtube.geos-chem.org).  These videos will demonstrate how to use the Github issue trackers and how to receive notifications about issues that have been posted:

  * [*Submitting GEOS-Chem issues on Github*](https://www.youtube.com/watch?v=dFBhdotYVf8&t=1103s)
  * [*Subscribing to Github notifications*](https://www.youtube.com/watch?v=RuH6zeYuzuY)
  * [*Searching for GEOS-Chem issues and pull requests*](https://www.youtube.com/watch?v=EiZC2vaXNnU&t=35s)

### Use our other issue templates for non-feature-request items

This issue template is intended for requesting new GEOS-Chem features and functionality.  Please use one of the other issue templates for:

* [Submitting a GEOS-Chem bug or technical issue](https://github.com/geoschem/geos-chem/issues/new?assignees=&labels=&template=report-a-bug-or-technical-issue.md&title=%5BBUG%2FISSUE%5D);
* [Asking a question about GEOS-Chem](https://github.com/geoschem/geos-chem/issues/new?assignees=&labels=&template=ask-a-question-about-geos-chem.md&title=%5BQUESTION%5D); and/or
* [Starting an extended GEOS-Chem discussion thread](https://github.com/geoschem/geos-chem/issues/new?assignees=&labels=discussion&template=start-a-geos-chem-discussion.md&title=%5BDISCUSSION%5D)

^^^^^^ YOU MAY DELETE ALL TEXT ABOVE THIS LINE WHEN CREATING YOUR ISSUE ^^^^^^

**Comment by jimmielin** [comment]
Hi Lizzie,

May I ask if the target for completion of this will be 13.0.0 or 12.9.0? In 13.0.0, the new HEMCO intermediate grid feature may shim the `HCO_EvalFld` call through a regrid. This will possibly implicate changing the `USE` statements. If 12.9.0 will finalize the work, I'll try not to change these calls until `dev/12.9.0` is merged into `dev/13.0.0` to avoid a bunch of merge conflicts.

Best,
Haipeng

**Comment by lizziel** [comment]
Hi Haipeng, 12.9.0 is basically wrapped up at this point. The parts of the update included in 12.9 are described in https://github.com/geoschem/geos-chem/issues/232. This new issue is meant to cover the remaining work, which may or may not go into 13.0. Don't worry much about USE statements. These are very easy and fast conflicts to fix.

**Comment by tsherwen** [comment]
Hello @lizziel,

I was wondering whether you were able to give an update on when this shift from "HCO_GetPtr" to "HCO_EvalFld" might be expected?

I am aware of a few of us still hardcoding scalings into the root modules in v12.9.1 (e.g. `strat_chem_mod.F90`). However being able to apply scalings via HEMCO would be simpler, faster and more versatile. 

Cheers, 

Tomas

**Comment by lizziel** [comment]
Hi @tsherwen, thanks for bringing this back up. `Strat_chem_mod.F90` is top priority in the list of files that are left with `Hco_GetPtr`. I was waiting for decisions to be made about the fate of trop and strat handling given talk of retiring tropchem. @msulprizio, what is the status on this? Is `strat_chem_mod.F90` now stable for the long-term?

**Comment by msulprizio** [comment]
Yes, `strat_chem_mod.F90` is stable as far as I know. It didn't change with the retiring of the "Tropchem" mechanism. Starting in 13.0.0, users will always use the "Standard" chemistry mechanism, now simply called "Fullchem." They will also choose the chemistry grid to be defined as the troposphere only or troposphere+stratosphere. The latter option turns on the `Input_Opt%UCX` switch in input.geos. When UCX is enabled, GMI prod/loss rates will be applied above the stratopause in `strat_chem_mod.F90` and when it is off UCX prod/loss rates will be applied above the tropopause in that module. This behavior is unchanged from previous versions. See https://github.com/geoschem/geos-chem/issues/332 for more details.

**Comment by lizziel** [comment]
Okay, great, thanks @msulprizio! Now it's a matter of finding time to do this update given we are trying to get 13.0 out as soon as possible. Realistically I think this will go into 13.1.

**Comment by jimmielin** [comment]
Thanks for the update @msulprizio and @lizziel, the plans on `strat_chem_mod.F90` will also coordinate well with the implementation of the intermediate HEMCO grid in GC-Classic for 13.1+

**Comment by tsherwen** [comment]
Thanks for the update @lizziel & @msulprizio!

**Comment by lizziel** [comment]
Update: I plan on revisiting the masking/scaling of all `HEMCO_Config.rc` entries following the 13.1 release.
